module/os/linux/zfs/zfs_uio.c

```diff
@@ -491,51 +496,106 @@ zfs_uio_free_dio_pages(zfs_uio_t *uio, zfs_uio_rw_t rw)
 
 		put_page(p);
 	}
-
+#endif
 	vmem_free(uio->uio_dio.pages,
 	    uio->uio_dio.npages * sizeof (struct page *));
 }
 
+#if defined(HAVE_PIN_USER_PAGES_UNLOCKED)
 static int
-zfs_uio_get_dio_pages_iov_iter(zfs_uio_t *uio, zfs_uio_rw_t rw)
+zfs_uio_pin_user_pages(zfs_uio_t *uio, zfs_uio_rw_t rw)
 {
+	long res;
 	size_t skip = uio->uio_skip;
+	size_t len = uio->uio_resid - skip;
+	unsigned int gup_flags = 0;
+	unsigned long addr;
+	unsigned long nr_pages;
+
+	/*
+	 * Kernel 6.2 introduced the FOLL_PCI_P2PDMA flag. This flag could
+	 * possibly be used here in the future to allow for P2P operations with
+	 * user pages.
+	 */
+	if (rw == UIO_READ)
+		gup_flags = FOLL_WRITE;
+
+	if (len == 0)
+		return (0);
+
+#if defined(HAVE_ITER_IS_UBUF)
+	if (iter_is_ubuf(uio->uio_iter)) {
+		nr_pages = DIV_ROUND_UP(len, PAGE_SIZE);
+		addr = (unsigned long)uio->uio_iter->ubuf + skip;
+		res = pin_user_pages_unlocked(addr, nr_pages,
+		    &uio->uio_dio.pages[uio->uio_dio.npages], gup_flags);
```

Comment:
I wonder if this code is supposed to handle page-misaligned buffers?  If yes, then I guess `nr_pages` might need correction for misaligned address, similar to what you seem to add below in `zfs_uio_get_dio_pages_iov_iter()`.  But then the `len != (res * PAGE_SIZE)` condition below will never be true.

Reply:
Good catch. I updated both call sites to update `npages` in this event.

Reply:
I was not actually arguing against accounting `start` in `zfs_uio_get_duio_upages_iov_iter()`, it seemed to make the code slightly more universal.  BTW, if you leave the assertion instead of what was there originally, I think it could be just an `ASSERT0(start)`.  I was thinking about the lack of it in `zfs_uio_pin_user_pages()`.  But whatever.